spec/spec_tests/uri_options_spec.rb

```diff
@@ -0,0 +1,123 @@
+require 'lite_spec_helper'
+
+describe 'Uri Options' do
+  include Mongo::ConnectionString
+
+  URI_OPTIONS_TESTS.each do |file|
+
+    spec = Mongo::ConnectionString::Spec.new(file)
+
+    context(spec.description) do
+
+      before(:all) do
+
+        module Mongo
+          class Address
+
+            private
+
+            alias :original_create_resolver :create_resolver
+            def create_resolver(timeout, ssl_options)
+              family = (host == 'localhost') ? ::Socket::AF_INET : ::Socket::AF_UNSPEC
+              info = ::Socket.getaddrinfo(host, nil, family, ::Socket::SOCK_STREAM)
+              FAMILY_MAP[info.first[4]].new(info[3], port, host)
+            end
+          end
+
+          class Server
+
+            # The constructor keeps the same API, but does not instantiate a
+            # monitor and run it.
+            alias :original_initialize :initialize
+            def initialize(address, cluster, monitoring, event_listeners, options = {})
+              @address = address
+              @cluster = cluster
+              @monitoring = monitoring
+              @options = options.freeze
+              @monitor = Monitor.new(address, event_listeners, Monitoring.new, options)
+            end
+
+            # Disconnect simply needs to return true since we have no monitor and
+            # no connection.
+            alias :original_disconnect! :disconnect!
+            def disconnect!; true; end
+          end
+        end
+      end
+
+      after(:all) do
+
+        module Mongo
+          # Return the implementations to their originals for the other
+          # tests in the suite.
+          class Address
+            alias :create_resolver :original_create_resolver
+            remove_method(:original_create_resolver)
+          end
+
+          class Server
+            alias :initialize :original_initialize
+            remove_method(:original_initialize)
+
+            alias :disconnect! :original_disconnect!
+            remove_method(:original_disconnect!)
+          end
+        end
+      end
+
+      spec.tests.each do |test|
+
+        # Skip optional tests for options the driver doesn't support.
+        if test.description.start_with?('tlsAllowInvalidHostnames') || test.description.start_with?('tlsInsecure')
```

Comment:
Why does Ruby not implement these options?

Reply:
According to the `Client` documentation, the `ssl_verify` option only toggles whether or not invalid certificates are accepted. Given that the driver currently does not implement the functionality to allow invalid hostnames, this option was left off (all of the other options added to this file already existed as `Client` options, just not in the URI).

The spec only requires that either `tlsInsecure` or one of `tlsAllowInvalidHostnames`/`tlsAllowInvalidCertificates` is implemented. Since we implement `tlsAllowInvalidCertificates`, we don't need to implement `tlsInsecure` (which would also currently be equivalent to `tlsAllowInvalidCertificates` since as mentioned above, the driver doesn't have the functionality to allow invalid hostnames.

Reply:
Ruby implemented hostname verification in https://github.com/ruby/openssl/pull/60. As far as I can tell this is included in 2.4.0. I cannot make out what the default is or when hostname verification gets turned on exactly, therefore, we should also have tests that the hostname verification is performed.

Reply:
I'm a little confused about what you're asking here. Are you asking that I go and implement hostname verification (which the driver currently doesn't do at all) as part of this PR? That feels a bit out of scope, given that it's not necessary work for the spec

Reply:
Since hostname verification is mentioned in the spec, that functionality should be implemented and documented in Ruby driver or there should be tickets created for this work.

Reply:
https://jira.mongodb.org/browse/RUBY-1608